code/modules/materials/_recipe.dm

```diff
@@ -0,0 +1,74 @@
+/*
+ * Recipe datum
+ */
+/datum/stack_recipe
+	var/title = "ERROR"
+	var/result_type
+	var/req_amount     //amount of material needed for this recipe
+	var/res_amount = 1 //amount of stuff that is produced in one batch (e.g. 4 for floor tiles)
+	var/max_res_amount = 1
+	var/time = 0
+	var/one_per_turf = 0
+	var/on_floor = 0
+	var/use_material
+	var/use_reinf_material
+	var/difficulty = 1 // higher difficulty requires higher skill level to make.
+	var/send_material_data = 0 //Whether the recipe will send the material name as an argument when creating product.
+	var/apply_material_name = 1 //Whether the recipe will prepend a material name to the title - 'steel clipboard' vs 'clipboard'
+
+#define CRAFTING_EXTRA_COST_FACTOR 1.2
+/datum/stack_recipe/New(material/material, var/reinforce_material)
+	if(material)
+		use_material = material.type
+		difficulty +=  material.construction_difficulty
+	if(reinforce_material)
+		use_reinf_material = reinforce_material
+	if(result_type && isnull(req_amount))
+		req_amount = 0
+		var/obj/I = new result_type
+		var/list/building_cost = I.building_cost()
+		for(var/path in building_cost)
+			req_amount += building_cost[path]
```

Comment:
Maybe worth adding a unit test that basically says: if you build a stack recipe, and then you retrieve its building cost, you don't wind up with (a) materials you didn't use in building it or (b) more of materials you used to build it than you started with. I don't think the logic here guarantees that such a test would always pass.

Reply:
Added!